File: AlgorithmsPython/SortingAlgorithms/Heapsort.py
'''
Created on May 3, 2020

@author: Sarthak Jain
'''


'''
Sorts the array using heap sort
'''
def heapSort(arr):
    # Sets n to the length of the array
    n = len(arr)
    # A loop that runs from half of n - 1 till -1, decrementing by 1
    for i in range(n // 2 - 1, -1, -1):
        # Goes into the heapify method, where it builds a MaxHeap
        heapify(arr, n, i)
    
    # A loop that runs from n-1 to 0, decrementing by 1
    for i in range(n - 1, 0, -1):
        
        # Swaps the value at index i and 0 in the array
        arr[i], arr[0] = arr[0], arr[i]
        
        # Goes into the heapify method where it builds a MaxHeap
        heapify(arr, i, 0)


'''
A helper method that builds a max heap
'''
def heapify(arr, n, i):
    # Sets largest to i
    largest = i
    
    # Sets left to 2 times i + 1
    left = 2 * i + 1
    
    # Sets right to 2 times i + 2
    right = 2 * i + 2
    
    # If left is smaller than n, and the value at index i is less than the value at index left, then it sets largest to left
    if left < n and arr[i] < arr[left]:
        largest = left
    
    # If right is smaller than n, and the value at index largest is less than the value at index right, then it sets largest to right    
    if right < n and arr[largest] < arr[right]:
        largest = right
    
    # If largest is not equal to i, then swaps the values at the largest and i index
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        
        # Recursively goes into the heapify method to make a MaxHeap
        heapify(arr, n, largest)

'''
Runs the main method where it sorts the array and prints out the sorted array
'''
if __name__ == "__main__": 
    print("Sorting an Array using HeapSort: ")
    arr = [24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 32435432532]
    print("Unsorted Array: " + str(arr));
    heapSort(arr)
    print("Sorted Array: " + str(arr))

File: AlgorithmsJava/src/sortingAlgorithms/HeapSort.java
package sortingAlgorithms;

import java.util.Arrays;

/**
 * A class that implements a generic HeapSort
 * 
 * @author Sarthak Jain
 *
 */
public class HeapSort {
	/**
	 * Runs the main method
	 */
	public static void main(String[] args) {
		Integer[] arr = new Integer[] { 24, 3476984, 11, 22, 69, 0, 88, 34, 331, 1, 23, 732, -1, 4325, 324352532 };
		System.out.println("Unsorted Array before HeapSort: " + Arrays.asList(arr));
		heapSort(arr);
		System.out.println("Sorted Array after HeapSort: " + Arrays.asList(arr));
	}

	/**
	 * Sorts a generic array using heapsort
	 */
	public static <T extends Comparable<? super T>> void heapSort(T[] arr) {

		// Sets length to array's length
		int length = arr.length;

		// Builds max heap in each iteration
		for (int i = length / 2 - 1; i >= 0; i--) {
			heapify(arr, length, i);
		}
		// Removes the top item from the heap and continues swapping, and recreating a
		// maxHeap
		for (int i = length - 1; i > 0; i--) {

			//Swaps the value at index 0 and index i 
			swap(arr, 0, i);

			//Creates a maxHeap
			heapify(arr, i, 0);
		}
	}

	/**
	 * Builds a generic max heap
	 */
	private static <T extends Comparable<? super T>> void heapify(T arr[], int n, int largeIndex) {
		// Initialize largest as largeIndex
		int largest = largeIndex;

		// Sets leftChild as 2*largeIndex + 1
		int leftChild = 2 * largeIndex + 1;

		// Sets rightChild as 2*largeIndex + 2
		int rightChild = 2 * largeIndex + 2;

		// If leftChild is smaller than n, and the value at index leftChild is bigger
		// than the value at index largest in the array, then set largest to leftChild
		if (leftChild < n && arr[leftChild].compareTo(arr[largest]) > 0)
			largest = leftChild;

		// If rightChild is smaller than n, and the value at index rightChild is bigger
		// than the value at index largest in the array, then set largest to rightChild
		if (rightChild < n && arr[rightChild].compareTo(arr[largest]) > 0)
			largest = rightChild;

		// If largest is not root
		if (largest != largeIndex) {

			// Swaps largeIndex and largest
			swap(arr, largeIndex, largest);

			// Recursively heapify the affected sub-tree
			heapify(arr, n, largest);
		}
	}

	/**
	 * A private helper method that does the swapping of the two numbers in the
	 * array
	 */
	private static <T> void swap(T[] arr, int firstNum, int secondNum) {
		// Gets a temp val based on the value at the firstNum in the array
		T temp = arr[firstNum];

		// Sets value at firstNum index to the value from the secondNum index in the arr
		arr[firstNum] = arr[secondNum];

		// Sets the value at the secondNum index to temp
		arr[secondNum] = temp;
	}

}
